src/Stratis.Bitcoin/Utilities/LockProtected.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>
+    /// Supports object-level locking and allows external work in the context of the locks.
+    /// Requires "lock (this.lockObject) { ...}" inside all of the object's public methods.
+    /// </summary>
+    public interface ILockProtected
+    {
+        /// <summary>Allows external work within lock context.</summary>
+        T Synchronous<T>(Func<T> action);
+
+        /// <summary>Allows external work within lock context.</summary>
+        void Synchronous(Action action);
+    }
+
+    public class LockProtected : ILockProtected
+    {
+        protected object lockObject { get; private set; }
+
+        public LockProtected()
+        {
+            this.lockObject = new object();
+        }
+
+        /// <inheritdoc />
+        public void Synchronous(Action action)
```

Comment:
Wondering if calling this `Locked` might be better? Or something similar?

Reply:
The name is chosen this way to indicate that the code will execute synchronously with other code of the object.